store/cachemulti/store.go

```diff
@@ -127,6 +127,32 @@ func (cms Store) Write() {
 	}
 }
 
+// Clone creates a snapshot of each store of the cache-multistore.
+// Each copy is a copy-on-write operation and therefore is very fast.
+func (cms Store) Clone() types.CacheMultiStore {
+	stores := make(map[types.StoreKey]types.CacheKVStore, len(cms.stores))
+	for key, store := range cms.stores {
+		stores[key] = store.Clone()
+	}
+	return Store{
+		db:     cms.db.Clone(),
+		stores: stores,
+
+		keys:         cms.keys,
+		traceWriter:  cms.traceWriter,
+		traceContext: cms.traceContext,
+	}
+}
+
+// Restore restores the cache-multistore cache to a given snapshot.
+func (cms Store) Restore(s types.CacheMultiStore) {
+	ms := s.(Store)
+	cms.db.Restore(ms.db)
+	for key, store := range cms.stores {
+		store.Restore(ms.stores[key])
+	}
+}
```

Comment:
- s.(Store) cannot be assumed to succeed
- cms.db.Restore mutates cms.db which requires taking a lock on cms
- ms.db cannot be accessed without acquiring a lock on ms
- The store value yielded by a range over cms.stores is a copy, calling Restore is not safe without acquiring a lock
- ms.stores[key] is not guaranteed to return a valid value

Reply:
> * s.(Store) cannot be assumed to succeed

This the the golang caveat I mentioned here: https://github.com/cosmos/cosmos-sdk/pull/14444#issuecomment-1397796711
Consider this as a contract of the API, also mentioned in the comments of the related interfaces.

> * cms.db.Restore mutates cms.db which requires taking a lock on cms

no, it don't mutate the reference, safety of `Restore` will be made sure by implementation, for example `cachekv.Store` has the locks internally.

> * ms.db cannot be accessed without acquiring a lock on ms

it's done inside of the `Restore`, the deadlock issue you mentioned, right?

> * The store value yielded by a range over cms.stores is a copy, calling Restore is not safe without acquiring a lock

- the store value is a reference
- the `stores` map don't mutate after construction, so it looks good to me.

> * ms.stores[key] is not guaranteed to return a valid value

that's a valid concern, I'll change it to a panic with a more friendlier message.

Reply:
> Consider this as a contract of the API, also mentioned in the comments of the related interfaces.

If this is the case, then why isn't `s` of type `Store`?

> safety of Restore will be made sure by implementation

Is this assumption documented? Do all implementations of Restore satisfy this expectation? Unless these things are true it cannot be assumed.

> the stores map don't mutate after construction

That's good and important, but it's also necessary that the values within that map are safe for concurrent use. Values in that map are interface types (CacheKVStore) and that interface does not guarantee that its implementations are safe for concurrent use. So your outer type must ensure synchronization somehow.